song2txt/txt_parser/metadata.py

```diff
@@ -0,0 +1,155 @@
+            text += model.format('GENRE', self.genre)
+        if self.year is not None:
+            text += model.format('YEAR', self.year)
+        if self.edition is not None:
+            text += model.format('EDITION', self.edition)
+        if self.creator is not None:
+            text += model.format('CREATOR', self.creator)
+        if self.language is not None:
+            text += model.format('LANGUAGE', self.language)
+        if self.cover is not None:
+            text += model.format('COVER', self.cover)
+        if self.vocals is not None:
+            text += model.format('VOCALS', self.vocals)
+        if self.video is not None:
+            text += model.format('VIDEO', self.video)
+        if self.background is not None:
+            text += model.format('BACKGROUND', self.background)
+        if self.relative is not None:
+            text += model.format('RELATIVE', 'YES' if self.relative else 'NO')
+        if self.gap is not None:
+            text += model.format('GAP', self._float_str(self.gap))
+        if self.video_gap is not None:
+            text += model.format('VIDEOGAP', self._float_str(self.video_gap))
+        if self.preview_start is not None:
+            text += model.format('PREVIEWSTART',
+                                 self._float_str(self.preview_start))
+        if self.name_singer_1 is not None:
+            text += model.format('DUETSINGERP1', self.name_singer_1)
+        if self.name_singer_2 is not None:
+            text += model.format('DUETSINGERP2', self.name_singer_2)
+        for key, value in self.extra.items():
+            text += model.format(key, value)
+
+        return text
+
+    def __eq__(self, other: object) -> bool:
+        return isinstance(other, MetaData) and self.__dict__ == other.__dict__
+
+    @staticmethod
+    def _float_str(num: float) -> str:
+        if num.is_integer():
+            return str(int(num))
+        else:
+            return str(num)
+
+    def get_bpm(self):
+        return self._bpm
+
+    def set_bpm(self, bpm):
+        if bpm > 0:
+            self._bpm = bpm
+        else:
+            raise ValueError('BPM must be greater than zero')
+
+    bpm = property(get_bpm, set_bpm)
+
+    def get_video_gap(self):
+        return self._video_gap
+
+    def set_video_gap(self, video_gap):
+        if video_gap is None or video_gap >= 0:
+            self._video_gap = video_gap
+        else:
+            raise ValueError('The video delay can not be negative')
+
+    video_gap = property(get_video_gap, set_video_gap)
+
+    def get_preview_start(self):
+        return self._preview_start
+
+    def set_preview_start(self, preview_start):
+        if preview_start is None or preview_start >= 0:
+            self._preview_start = preview_start
+        else:
+            raise ValueError('Second for preview start can not be negative')
+
+    preview_start = property(get_preview_start, set_preview_start)
+
+    def add(self, key: str, value: str):
```

Comment:
Uf, so much boilerplate...
También, setter sin comprobación de tipos... también asignas bpm sin comprobar nada y sin usar el setter que ya tienes... En general, este tipo de cosas llevan a complicaciones e indican que no está tan claro cómo se va a usar la clase _realmente_ De veras, pégale otra vuelta y crea issues que planteen los problemas que hay.

Reply:
Si se hace la comprobación en el constructor! Las funciones no se usan, es una sintaxis especial de python para cambiar el comportamiento por defecto de los atributos (https://stackoverflow.com/questions/2627002/whats-the-pythonic-way-to-use-getters-and-setters), esto se aplica también en el constructor y en cualquier uso del atributo en general.

Respecto a la cantidad de código de add, que fundamentalmente es para saber a que atributo corresponde cada TAG, es cierto que es bastante horrible, larga y repetitiva. Sinceramente a lo único que llego para simplificarlo es hacer un diccionario con los nombres de los atributos relacionándolo con los TAGS, aunque la veo algo fea. No se que te parece esa solución o si se te ocurre alguna otra en concreto.

Reply:
Agh, eso no lo sabía.
Lo del diccionario es una opción, pero creo recordar que se puede acceder a los atributos por nombre usando `Object`, ¿puede ser?

Reply:
Sí, no recuerdo exactamente como pero estoy seguro que se puede acceder a los atributos por nombre. La idea era las TAGS (lo que sale en el txt para indicar el tipo de metadato) saber a que nombre de atributo corresponden. También podría darles el mismo nombre que tienen las tags pero en minúscula y no haría falta diccionario ahora que lo pienso, aunque me saltaría la guía de estilo al juntar varias palabras sin separarlas con _